common/thorhelper/thorcommon.cpp

```diff
@@ -976,141 +986,231 @@ class CRowStreamReader : implements IExtRowStream, public CSimpleInterface
         }
     } crccb;
     
-public:
-    IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+    inline bool fieldFilterMatch(const void * buffer)
+    {
+        if (actualFilter.numFilterFields())
+        {
+            filterRow->setRow(buffer, 0);
+            return actualFilter.matches(*filterRow);
+        }
+        else
+            return true;
+    }
 
-    CRowStreamReader(IFileIO *_fileio, IMemoryMappedFile *_mmfile, IRowInterfaces *rowif, offset_t _ofs, offset_t _len, bool _tallycrc, EmptyRowSemantics _emptyRowSemantics)
-        : fileio(_fileio), mmfile(_mmfile), allocator(rowif->queryRowAllocator()), prefetchBuffer(NULL), emptyRowSemantics(_emptyRowSemantics)
+    inline bool checkEmptyRow()
+    {
+        if (ers_allow == emptyRowSemantics)
+        {
+            byte b;
+            prefetchBuffer.read(1, &b);
+            prefetchBuffer.finishedRow();
+            if (1 == b)
+                return true;
+        }
+        return false;
+    }
+    inline void checkEog()
+    {
+        if (ers_eogonly == emptyRowSemantics)
+        {
+            byte b;
+            prefetchBuffer.read(1, &b);
+            eog = 1 == b;
+        }
+    }
+    inline bool checkExitConditions()
+    {
+        if (prefetchBuffer.eos())
+        {
+            eos = true;
+            return true;
+        }
+        if (eog)
+        {
+            eog = false;
+            if (hadMatchInGroup)
+            {
+                hadMatchInGroup = false;
+                return true;
+            }
+        }
+        return false;
+    }
+    const byte *getNextPrefetchRow()
+    {
+        while (true)
+        {
+            ++progress;
+            if (checkEmptyRow())
+                return nullptr;
+            currentRowOffset = prefetchBuffer.tell();
+            prefetcher->readAhead(prefetchBuffer);
+            const byte * row = prefetchBuffer.queryRow();
+            bool matched = fieldFilterMatch(row);
+            checkEog();
+            if (matched) // NB: prefetchDone() call must be paired with a row returned from prefetchRow()
+            {
+                hadMatchInGroup = true;
+                return row;
+            }
+            else
+                prefetchBuffer.finishedRow();
+            if (checkExitConditions())
```

Comment:
This logic needs to be at the head of the loop - otherwise a row at the end of group, that is returned will not return eog after it.  Need some more tests that ensure that the grouping is also correctly preserved.

Reply:
There is code for eog in the caller - but that isn't correct because it doesn't check/reset hadMatchInGroup

Reply:
yes, hadMatchInGroup should be have been reset.
I extended the regression test to expose the problem and confirmed the fix once hadMatchInGroup reset.